### PR TITLE
add XQuery test to check for missing entries in language files

### PR DESCRIPTION
**Pull request by peterstadler**
## Description, Context and related Issue

This PR introduces a new XQuery based check for missing entries in language files

Refs #154

## How Has This Been Tested?

Checked locally by modifying language files and checking results of running `ant check-language-files`

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist

- I have read the [CONTRIBUTING](https://github.com/Edirom/Edirom-Online/blob/develop/CONTRIBUTING.md) document.

### CODE REVIEW

**Comment by peterstadler** [comment]
I've also added the new check "check-language-files" to the Github actions:
1. The run https://github.com/Edirom/Edirom-Online/actions/runs/11352963364/job/31576942488 failed due to a missing entry in the German language file
2. After adding the missing entry the Github action passed: https://github.com/Edirom/Edirom-Online/actions/runs/11353084552/job/31577350564

**Comment by roewenstrunk** [comment]
When I run the test, I get an error because I don't have the right libraries for ant: `[java] Could not find net.sf.saxon.Query. Make sure you have it in your classpath`.
Would you like to write a minimal description in the [wiki](https://github.com/Edirom/Edirom-Online/wiki) of how this type of tests work and what requirements must be met?

**Comment by peterstadler** [comment]
I have added some basic information regarding dependencies in the Readme: https://github.com/Edirom/Edirom-Online/blob/3e57257bc5d1f8d095c715c7c93cfc1282c7af85/README.md?plain=1#L104-L110

**Comment by peterstadler** [comment]
pinging @bwbohl and @roewenstrunk for reviewing

**Comment by bwbohl** [comment]
Ok, I successfully tested the lang-test by doing the following:

1. Use [bwbohl/sencha-cmd](https://github.com/bwbohl/sencha-cmd/pkgs/container/sencha-cmd) as described in the [README](https://github.com/Edirom/Edirom-Online?tab=readme-ov-file#building-locally).

2. Install missing dependecies for the tests

   ```
   apt-get update
   ```
 
   ```
   apt-get install libsaxonhe-java
   ```
   
 3. from within '/app' run the test, referencing the JARs and the buildfile
 
 ```
 ant -lib /usr/share/java/ -buildfile testing/ant-tests.xml check-language-files
 ```
 
This reported nothing and when changing the name of one entry it reported the original term missing in the modified file and the modified term in the other file. Everything as expected, very nice!

**Review by bwbohl — APPROVED** [review]
Just a short follow-up: Should we also check `add/data/xslt/iq8n`?

**Comment by bwbohl** [comment]
Thanks for imrproving!
Lot’s of discrepancies surfaced this way. Probably we should fix. them before merging ;-)

**Comment by bwbohl** [comment]
Just trying to fix the files and realised maybe we should also cover duplicates in the error report. Currently this would be the result for a duplicate entry in one file:
```xml
<key missing="" available="en de">csl</key>
```

**Comment by peterstadler** [comment]
Regarding duplicates: Shall I add proper output for that case or is the current (undocumented) way of surfacing duplicates ok for now?

**Comment by bwbohl** [comment]
I’d prefer making it clean now, so e.g.

```xml
<key missing="" available="en de" duplicate="de">csl</key>
```

File: add/data/xslt/i18n/en.xml [hunk 1]
```diff
@@ -119,25 +152,32 @@
     <entry xml:id="settlement">Settlement</entry>
     <entry xml:id="supportDesc">Support Description</entry>
     <entry xml:id="handDesc">Description of Hands</entry>
+    <entry xml:id="handNotes">Notes on hands</entry>
     <entry xml:id="itemList">Item List</entry>
-    <entry xml:id="pubStmt">Publication Statement</entry>
+    <entry xml:id="placeName">Ort</entry>
```
Comment (peterstadler) [inline add/data/xslt/i18n/en.xml:157]: Really German? Or should that rather be an English term?

File: add/data/xslt/i18n/en.xml [hunk 2]
```diff
@@ -119,25 +152,32 @@
...
+    
+    <entry xml:id="title_desc">Descriptive title</entry>
+    <entry xml:id="title_main">Main title</entry>
+    <entry xml:id="title_sub">Subordinate title</entry>
+    
     <!-- FreiDi /content -->
     <entry xml:id="width">width</entry>
     <entry xml:id="height">height</entry>
     <entry xml:id="descLink">Source Description</entry>
+    <entry xml:id="singers">Sänger</entry>
```
Comment (peterstadler) [inline add/data/xslt/i18n/en.xml:181]: Really German? Or should that rather be an English term?

**Comment by peterstadler** [comment]
added explicit warning for duplicate keys. Output looks like:
```xml
<results>
   <key duplicates="de en">controller.window.Window_renderingView</key>
   <key missing="en" available="de">global_execute</key>
   <key duplicates="de">controller.window.Window_summaryView</key>
   <key missing="de" available="en">global_execue</key>
</results>
```

File: add/data/xslt/i18n/en.xml (hunk 2 above)
Comment (bwbohl) [reply add/data/xslt/i18n/en.xml:181]:
```suggestion
    <entry xml:id="singers">Singers</entry>
```

File: add/data/xslt/i18n/en.xml (hunk 1 above)
Comment (bwbohl) [reply add/data/xslt/i18n/en.xml:157]:
```suggestion
    <entry xml:id="placeName">Place name</entry>
```

**Comment by bwbohl** [comment]
After retesting I have to report that it works fine for the files in `locale` but does not report duplicates in the new style when there are some in the ì18n` files.

Moreover if there˚s a duplicate in one of the `locale` files and another in the ì18n` files only the `locale` gets reported.

File: testing/check-language-files.xq [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        where count($entry) ne count($language-files)
+        let $available :=
+        (
+            $entry/ancestor::langFile/lang ! string(.), (: language files in add/data/locale :)
+            $entry/parent::language/@xml:lang ! string(.) (: language files in add/data/xslt/i18n :)
+        )
+        let $missing := $languages[not(. = $available)]
+        let $duplicates :=
+            for $i in $entry
+            group by $langkey := $i/ancestor::langFile/lang
```
Comment (bwbohl) [inline testing/check-language-files.xq:43]: fix duplicates test for ì18n` files

```suggestion
            group by $langkey := ($i/ancestor::langFile/lang, $i/parent::language/@xml:lang)
```

File: testing/ant-tests.xml [hunk 4]
```diff
@@ -48,5 +48,43 @@
...
+                Missing entries in language files at /add/data/xslt/i18n: &#10;${xquery.result}
+                <condition>
+                    <!--
+                        If all goes well, "<results/>" will be returned of string-length=10
+                        Any errors reported will add to this string-length hence cause the test to fail
+                    -->
+                    <length string="${xquery.result}" when="greater" length="10"/>
+                </condition>
+            </fail>
+        </sequential>
```
Comment (bwbohl) [inline testing/ant-tests.xml:87]: Execute both tests in parallel to make sure both report their output

```suggestion
        <parallel>
            <sequential>
                <local name="xquery.result"/>
                <java classname="net.sf.saxon.Query" failonerror="false" outputproperty="xquery.result">
                    <arg value="testing/check-language-files.xq"/>
                    <arg value="{http://www.edirom.de/ns/1.3}language-files-path=../add/data/locale?select=*.xml"/>
                </java>
                <fail>
                    Missing entries in language files at /add/data/locale: &#10;${xquery.result}
                    <condition>
                        <!-- 
                            If all goes well, "<results/>" will be returned of string-length=10
                            Any errors reported will add to this string-length hence cause the test to fail
                        -->
                        <length string="${xquery.result}" when="greater" length="10"/>
                    </condition>
                </fail>
            </sequential>
            <sequential>
                <local name="xquery.result"/>
                <java classname="net.sf.saxon.Query" failonerror="false" outputproperty="xquery.result">
                    <arg value="testing/check-language-files.xq"/>
                    <arg value="{http://www.edirom.de/ns/1.3}language-files-path=../add/data/xslt/i18n?select=*.xml"/>
                </java>
                <fail>
                    Missing entries in language files at /add/data/xslt/i18n: &#10;${xquery.result}
                    <condition>
                        <!--
                            If all goes well, "<results/>" will be returned of string-length=10
                            Any errors reported will add to this string-length hence cause the test to fail
                        -->
                        <length string="${xquery.result}" when="greater" length="10"/>
                    </condition>
                </fail>
            </sequential>
        </parallel>
```

**Comment by peterstadler** [comment]
Thank you very much @bwbohl for the thorough review! 
In added your suggestions (great idea to run the tests in parallel!) and finally did some very explicit typing ;) 

Works on my machine but it'd be great if you could have yet another (last) look.

**Review by bwbohl — APPROVED** [review]
Added some minor code linting.
Works like a charm, great implementation work, love it!

**Pull request merged**